hooks/basic/collector.py

```diff
@@ -207,18 +207,76 @@ def on_context_changed(self, settings, item):
+
+
+    def _resolve_work_path_template(self, properties, path):
+        """
+        Resolve work_path_template from the properties.
+        The signature uses properties, so that it can resolve the template even if the item object hasn't been created.
+
+        :param properties: properties that have/will be used to build item object.
+        :param path: path to be used to get the templates, using template_from_path.
+        :return: Name of the template.
+        """
+
+        work_path_template = properties.get("work_path_template")
 
         # If defined, add the work_path_template to the item's properties
+        # set/evaluate the correct work_path_template
+        # First check if a work_path_template has been defined
         if work_path_template:
-            item.properties["work_path_template"] = work_path_template
+            work_tmpl = self.parent.get_template_by_name(work_path_template)
+            if not work_tmpl:
+                # this template was not found in the template config!
+                raise TankError("The template '%s' does not exist!" % work_path_template)
 
-        # Set the item's fields property
-        item.properties["fields"] = self._resolve_item_fields(item)
+            # update the field with correct value so that we can use it everytime for this item
+            work_path_template = work_tmpl.name
+
+        # Else see if the path matches an existing template
+        else:
+            # let's try to check if this path fits into any known template
+            work_tmpl = self.sgtk.template_from_path(path)
+            if not work_tmpl:
+                # this path doesn't map to any known templates!
+                self.logger.warning("Cannot find a matching template for path: %s" % path)
+            else:
+                # update the field with correct value so that we can use it everytime for this item
+                work_path_template = work_tmpl.name
+
+        return work_path_template
+
+
+    def _get_item_context_from_path(self, properties, path):
+        """Updates the context of the item from the work_path_template/template, if needed.
+
+        :param properties: properties of the item.
+        :param path: path to build the context from.
+        """
+
+        publisher = self.parent
+        new_context = None
+
+        work_path_template = self._resolve_work_path_template(properties, path)
+
+        if work_path_template:
+            work_tmpl = publisher.get_template_by_name(work_path_template)
+            # evaluate the context of the item too at the time of addition.
+            # user can change this later if they don't like it.
+            # this is being done here, since the tasks are created after the process_file hook runs.
+            # We anyways can't resolve a TASK context using context_from_path?!
+            # So we can check for that? In case we are launching the publisher with correct context.
+            if not publisher.context.task:
```

Comment:
This is evaluating the application context, instead it should be evaluating the item's (or rather, parent_item's) context. Also, a non Task-based context is valid, so we shouldn't be overwriting it based on whether a Task is defined... Maybe the better option is to compare the two contexts. Try this...

```
new_context = self.tank.context_from_entities(entities, parent_item.context)
if new_context != parent_item.context:
    return new_context
else:
    return parent_item.context
```

That functionality should be available in `sgtk_core-2.11.1`

Reply:
Since item.context would return _bundle.context. I was just looking for a way to distinguish when to update the context of item. parent_item.context is a better way to write it. I will make the changes.

The idea behind checking for non-task based contexts was that, when you launch the publisher from a DCC application, the user can only do that after they have chosen a task and set the context. That case was only to make sure that we don't update the item with a wrong context. Since if when ran context_from_entities previously it won't return a context with a task. With the updated context_from_entities method it would.